File: 1.x/demos/iOS/Example-iOS-ObjC/Example-iOS-ObjC/UIUtils.m
//
//  UIUtils.m
//  Example-iOS-ObjC
//
//  Created by Hildequias Junior on 8/8/17.
//  Copyright © 2017 UOL Pagseguro. All rights reserved.
//

#import "UIUtils.h"
#import "PlugPag.h"
#import "SVProgressHUD.h"

@implementation UIUtils

+ (NSMutableAttributedString *) formatTextGetLastApprovedTransaction {
    
    NSString *txMensagem = @"Mensagem:";
    NSString *txTrCode = @"Tr. Code:";
    NSString *txDate = @"Date:";
    NSString *txTime = @"Time:";
    NSString *txHost = @"Host NSU:";
    NSString *txCardBrand = @"Card Brand:";
    NSString *txBin = @"Bin:";
    NSString *txHolder = @"Holder:";
    NSString *txUserReference = @"User Reference:";
    NSString *txTerSerial = @"Terminal Serial:";
    
    NSString *strTextView = [NSString stringWithFormat:@"%@ %@ \n%@ %@ \n%@ %@ \n%@ %@ \n%@ %@ \n%@ %@ \n%@ %@ \n%@ %@ \n%@ %@ \n%@ %@ ",        txMensagem, [PlugPag sharedInstance].message,
                             txTrCode, [PlugPag sharedInstance].transactionCode,
                             txDate, [PlugPag sharedInstance].date,
                             txTime, [PlugPag sharedInstance].time,
                             txHost, [PlugPag sharedInstance].hostNsu,
                             txCardBrand, [PlugPag sharedInstance].cardBrand,
                             txBin, [PlugPag sharedInstance].bin,
                             txHolder, [PlugPag sharedInstance].holder,
                             txUserReference, [PlugPag sharedInstance].userReference,
                             txTerSerial, [PlugPag sharedInstance].terminalSerialNumber];
    
    NSRange rangeMensagem       = [strTextView rangeOfString:txMensagem];
    NSRange rangeTrCode         = [strTextView rangeOfString:txTrCode];
    NSRange rangeDate           = [strTextView rangeOfString:txDate];
    NSRange rangeTime           = [strTextView rangeOfString:txTime];
    NSRange rangeHost           = [strTextView rangeOfString:txHost];
    NSRange rangeCardBrand      = [strTextView rangeOfString:txCardBrand];
    NSRange rangeBin            = [strTextView rangeOfString:txBin];
    NSRange rangeHolder         = [strTextView rangeOfString:txHolder];
    NSRange rangeUserReference  = [strTextView rangeOfString:txUserReference];
    NSRange rangeTerSerial      = [strTextView rangeOfString:txTerSerial];
    
    UIFont *fontText = [UIFont boldSystemFontOfSize:16];
    NSDictionary *dictBoldText = [NSDictionary dictionaryWithObjectsAndKeys:fontText, NSFontAttributeName, nil];
    
    NSMutableAttributedString *mutAttrTextViewString = [[NSMutableAttributedString alloc] initWithString:strTextView];
    [mutAttrTextViewString setAttributes:dictBoldText range:rangeMensagem];
    [mutAttrTextViewString setAttributes:dictBoldText range:rangeTrCode];
    [mutAttrTextViewString setAttributes:dictBoldText range:rangeDate];
    [mutAttrTextViewString setAttributes:dictBoldText range:rangeTime];
    [mutAttrTextViewString setAttributes:dictBoldText range:rangeHost];
    [mutAttrTextViewString setAttributes:dictBoldText range:rangeCardBrand];
    [mutAttrTextViewString setAttributes:dictBoldText range:rangeBin];
    [mutAttrTextViewString setAttributes:dictBoldText range:rangeHolder];
    [mutAttrTextViewString setAttributes:dictBoldText range:rangeUserReference];
    [mutAttrTextViewString setAttributes:dictBoldText range:rangeTerSerial];
    
    return mutAttrTextViewString;
}

+ (NSMutableAttributedString *) formatTextError:(int) error withMessage:(NSString *) msg {
    
    NSString *message = [NSString stringWithFormat:@"Erro: %d - %@", error, msg];
    
    NSRange range = [message rangeOfString:message];
    UIFont *fontText = [UIFont systemFontOfSize:19];
    NSMutableParagraphStyle *style = [NSMutableParagraphStyle new];
    style.alignment = NSTextAlignmentCenter;
    
    NSDictionary *dictText = [NSDictionary dictionaryWithObjectsAndKeys:fontText, NSFontAttributeName,
                              [UIColor blackColor], NSForegroundColorAttributeName,
                              style, NSParagraphStyleAttributeName,
                              nil];
    
    NSMutableAttributedString *mutAttrTextViewString = [[NSMutableAttributedString alloc] initWithString:message];
    [mutAttrTextViewString setAttributes:dictText range:range];
    
    return mutAttrTextViewString;
}

+ (NSMutableAttributedString *) formatTextResult:(NSString *) message {
    
    NSRange range = [message rangeOfString:message];
    UIFont *fontText = [UIFont systemFontOfSize:19];
    NSMutableParagraphStyle *style = [NSMutableParagraphStyle new];
    style.alignment = NSTextAlignmentCenter;
    
    NSDictionary *dictText = [NSDictionary dictionaryWithObjectsAndKeys:fontText, NSFontAttributeName,
                              [UIColor blackColor], NSForegroundColorAttributeName,
                              style, NSParagraphStyleAttributeName,
                              nil];
    
    NSMutableAttributedString *mutAttrTextViewString = [[NSMutableAttributedString alloc] initWithString:message];
    [mutAttrTextViewString setAttributes:dictText range:range];
    
    return mutAttrTextViewString;
}

+ (NSMutableAttributedString *)formatCurrency:(float) value {
    
    NSNumberFormatter *n = [[NSNumberFormatter alloc] init];
    [n setNumberStyle:NSNumberFormatterCurrencyStyle];
    NSLocale *locale = [[NSLocale alloc] initWithLocaleIdentifier:@"pt_BR"];
    [n setLocale:locale];
    NSString *message = [n stringFromNumber:[NSNumber numberWithFloat:value]];
    
    NSRange range    = [message rangeOfString:message];
    UIFont *fontText = [UIFont boldSystemFontOfSize:20.0];
    NSMutableParagraphStyle *style = [NSMutableParagraphStyle new];
    style.alignment  = NSTextAlignmentCenter;
    
    NSDictionary *dictText = [NSDictionary dictionaryWithObjectsAndKeys:fontText, NSFontAttributeName,
                              [UIColor blackColor], NSForegroundColorAttributeName,
                              style, NSParagraphStyleAttributeName,
                              nil];
    
    NSMutableAttributedString *mutAttrTextViewString = [[NSMutableAttributedString alloc] initWithString:message];
    [mutAttrTextViewString setAttributes:dictText range:range];
    
    return mutAttrTextViewString;
}

+ (void) showProgress {
    
    [SVProgressHUD show];
}

+ (void) hideProgress {
    
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        // time-consuming task
        dispatch_async(dispatch_get_main_queue(), ^{
            [SVProgressHUD dismiss];
        });
    });
}

+ (void) showAlert:(UIViewController *) view withMessage:(NSString *) message {
    
    UIAlertController * alert=[UIAlertController alertControllerWithTitle:@"Aviso" message:message preferredStyle:UIAlertControllerStyleAlert];
    
    UIAlertAction* yesButton = [UIAlertAction
                                actionWithTitle:@"Ok"
                                style:UIAlertActionStyleDefault
                                handler:^(UIAlertAction * action)
                                {}];
    
    [alert addAction:yesButton];
    
    [view presentViewController:alert animated:YES completion:nil];
}



@end

File: 1.x/demos/iOS/Example-iOS-Swift/Example-iOS-Swift/UIUtils.swift
//
//  UIUtils.swift
//  Example-iOS-Swift
//
//  Created by Hildequias Junior on 8/22/17.
//  Copyright © 2017 UOL Pagseguro. All rights reserved.
//

import UIKit

extension String {
    func nsRange(from range: Range<String.Index>) -> NSRange {
        let from = range.lowerBound.samePosition(in: utf16)
        let to = range.upperBound.samePosition(in: utf16)
        return NSRange(location: utf16.distance(from: utf16.startIndex, to: from),
                       length: utf16.distance(from: from, to: to))
    }
}

extension String {
    func range(from nsRange: NSRange) -> Range<String.Index>? {
        guard
            let from16 = utf16.index(utf16.startIndex, offsetBy: nsRange.location, limitedBy: utf16.endIndex),
            let to16 = utf16.index(utf16.startIndex, offsetBy: nsRange.location + nsRange.length, limitedBy: utf16.endIndex),
            let from = from16.samePosition(in: self),
            let to = to16.samePosition(in: self)
            else { return nil }
        return from ..< to
    }
}

class UIUtils: NSObject {
    
    class func formatTextGetLastApprovedTransaction() -> NSMutableAttributedString {
        
        let txMensagem = "Mensagem:"
        let txTrCode = "Tr. Code:"
        let txDate = "Date:"
        let txTime = "Time:"
        let txHost = "Host NSU:"
        let txCardBrand = "Card Brand:"
        let txBin = "Bin:"
        let txHolder = "Holder:"
        let txUserReference = "User Reference:"
        let txTerSerial = "Terminal Serial:"
        
        let strTextView = String("\(txMensagem) \(PlugPag.sharedInstance().message!) \n \(txTrCode) \(PlugPag.sharedInstance().transactionCode!) \n \(txDate) \(PlugPag.sharedInstance().date!) \n \(txTime) \(PlugPag.sharedInstance().time!) \n \(txHost) \(PlugPag.sharedInstance().hostNsu!) \n \(txCardBrand) \(PlugPag.sharedInstance().cardBrand!) \n \(txBin) \(PlugPag.sharedInstance().bin!) \n \(txHolder) \(PlugPag.sharedInstance().holder!) \n \(txUserReference) \(PlugPag.sharedInstance().userReference!) \n \(txTerSerial) \(PlugPag.sharedInstance().terminalSerialNumber!)")
        
        let rangeMensagem           = strTextView?.range(of: txMensagem)
        let rangeTrCode             = strTextView?.range(of: txTrCode)
        let rangeDate               = strTextView?.range(of: txDate)
        let rangeTime               = strTextView?.range(of: txTime)
        let rangeHost               = strTextView?.range(of: txHost)
        let rangeCardBrand          = strTextView?.range(of: txCardBrand)
        let rangeBin                = strTextView?.range(of: txBin)
        let rangeHolder             = strTextView?.range(of: txHolder)
        let rangeUserReference      = strTextView?.range(of: txUserReference)
        let rangeTerSerial          = strTextView?.range(of: txTerSerial)
        
        let fontText = UIFont.boldSystemFont(ofSize: 16.0)
        
        let dictBoldText: [String : Any] = [
            NSForegroundColorAttributeName : UIColor.black,
            NSFontAttributeName : fontText
        ]
        
        let mutAttrTextViewString = NSMutableAttributedString(string: strTextView!)
        mutAttrTextViewString.setAttributes(dictBoldText, range: (strTextView?.nsRange(from: rangeMensagem!))!)
        mutAttrTextViewString.setAttributes(dictBoldText, range: (strTextView?.nsRange(from: rangeTrCode!))!)
        mutAttrTextViewString.setAttributes(dictBoldText, range: (strTextView?.nsRange(from: rangeDate!))!)
        mutAttrTextViewString.setAttributes(dictBoldText, range: (strTextView?.nsRange(from: rangeTime!))!)
        mutAttrTextViewString.setAttributes(dictBoldText, range: (strTextView?.nsRange(from: rangeHost!))!)
        mutAttrTextViewString.setAttributes(dictBoldText, range: (strTextView?.nsRange(from: rangeCardBrand!))!)
        mutAttrTextViewString.setAttributes(dictBoldText, range: (strTextView?.nsRange(from: rangeBin!))!)
        mutAttrTextViewString.setAttributes(dictBoldText, range: (strTextView?.nsRange(from: rangeHolder!))!)
        mutAttrTextViewString.setAttributes(dictBoldText, range: (strTextView?.nsRange(from: rangeUserReference!))!)
        mutAttrTextViewString.setAttributes(dictBoldText, range: (strTextView?.nsRange(from: rangeTerSerial!))!)
        
        return mutAttrTextViewString;
    }
    
    class func formatTextError(error: Int, msg: String) -> NSMutableAttributedString {
        
        let message = "Erro: \(error) - \(msg)"
        
        let range = message.range(of: message)
        let font = UIFont.systemFont(ofSize: 19.0)
        let styleMsg = NSMutableParagraphStyle()
        styleMsg.alignment = NSTextAlignment.center
        
        let dictText: [String : Any] = [
            NSForegroundColorAttributeName : UIColor.black,
            NSParagraphStyleAttributeName: styleMsg,
            NSFontAttributeName : font
        ]
        
        let txErro = "Erro:"
        let rangeErro = message.range(of: txErro)
        let fontErro = UIFont.boldSystemFont(ofSize: 20.0)
        let styleErro = NSMutableParagraphStyle()
        styleErro.alignment = NSTextAlignment.center
        
        let dictBoldText: [String : Any] = [
            NSForegroundColorAttributeName : UIColor.black,
            NSParagraphStyleAttributeName: styleErro,
            NSFontAttributeName : fontErro
        ]
        
        let mutAttrTextViewString = NSMutableAttributedString(string: message)
        mutAttrTextViewString.setAttributes(dictText, range: (message.nsRange(from: range!)))
        mutAttrTextViewString.setAttributes(dictBoldText, range: (message.nsRange(from: rangeErro!)))
        
        return mutAttrTextViewString;
    }
    
    class func formatTextResult(message: String) -> NSMutableAttributedString {
        
        let range = message.range(of: message)
        let font = UIFont.systemFont(ofSize: 19.0)
        let styleMsg = NSMutableParagraphStyle()
        styleMsg.alignment = NSTextAlignment.center
        
        let dictText: [String : Any] = [
            NSForegroundColorAttributeName : UIColor.black,
            NSParagraphStyleAttributeName: styleMsg,
            NSFontAttributeName : font
        ]
        
        let mutAttrTextViewString = NSMutableAttributedString(string: message)
        mutAttrTextViewString.setAttributes(dictText, range: (message.nsRange(from: range!)))
        
        return mutAttrTextViewString;
    }
    
    class func formatCurrency(value: Float) -> NSMutableAttributedString {
        
        let formatter = NumberFormatter()
        formatter.locale = NSLocale(localeIdentifier: "pt_BR") as Locale!
        formatter.numberStyle = .currency
        
        let msg = formatter.string(from: value as NSNumber)!
        
        let range = msg.range(of: msg)
        let fontText = UIFont.boldSystemFont(ofSize: 20.0)
        let style = NSMutableParagraphStyle()
        style.alignment = NSTextAlignment.center
        
        let dictBoldText: [String : Any] = [
            NSForegroundColorAttributeName : UIColor.black,
            NSParagraphStyleAttributeName: style,
            NSFontAttributeName : fontText
        ]
        
        let mutAttrTextViewString = NSMutableAttributedString(string: msg)
        mutAttrTextViewString.setAttributes(dictBoldText, range: (msg.nsRange(from: range!)))
        
        return mutAttrTextViewString;
    }
    
    class func showProgress () {
        
        SVProgressHUD.show()
    }
    
    class func hideProgress () {
        
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
            SVProgressHUD.dismiss()
        }
    }
    
    class func showAlert (view: UIViewController, message: String) {
        
        let alert=UIAlertController(title: "Aviso", message: message, preferredStyle: UIAlertControllerStyle.alert)
        
        alert.addAction(UIAlertAction(title: "Ok", style: UIAlertActionStyle.default, handler: {(action:UIAlertAction) in }))
        view.present(alert, animated: true, completion: nil)
    }
    
}
